src/util/pointer_expr.h

```diff
@@ -381,6 +381,235 @@ inline address_of_exprt &to_address_of_expr(exprt &expr)
+///
+/// \param expr: Source expression
+/// \return Object of type \ref object_address_exprt
+inline const object_address_exprt &to_object_address_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_object_address);
+  const object_address_exprt &ret =
+    static_cast<const object_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \copydoc to_object_address_expr(const exprt &)
+inline object_address_exprt &to_object_address_expr(exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_object_address);
+  object_address_exprt &ret = static_cast<object_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \brief Operator to return the address of a field relative
+/// to a base address
+class field_address_exprt : public unary_exprt
+{
+public:
+  /// constructor for field addresses.
+  /// The base address must be a pointer to a compound type.
+  field_address_exprt(
+    exprt base,
+    const irep_idt &component_name,
+    pointer_typet type);
+
+  exprt &base()
+  {
+    return op0();
+  }
+
+  const exprt &base() const
+  {
+    return op0();
+  }
+
+  const pointer_typet &type() const
+  {
+    return static_cast<const pointer_typet &>(exprt::type());
+  }
+
+  pointer_typet &type()
+  {
+    return static_cast<pointer_typet &>(exprt::type());
+  }
+
+  /// returns the type of the field whose address is represented
+  const typet &field_type() const
+  {
+    return type().subtype();
+  }
+
+  const typet &compound_type() const
+  {
+    return to_pointer_type(base().type()).subtype();
+  }
+
+  const irep_idt &component_name() const
+  {
+    return get(ID_component_name);
+  }
+};
+
+template <>
+inline bool can_cast_expr<field_address_exprt>(const exprt &expr)
+{
+  return expr.id() == ID_field_address;
+}
+
+inline void validate_expr(const field_address_exprt &value)
+{
+  validate_operands(value, 1, "field_address must have one operand");
+}
+
+/// \brief Cast an exprt to an \ref field_address_exprt
+///
+/// \a expr must be known to be \ref field_address_exprt.
+///
+/// \param expr: Source expression
+/// \return Object of type \ref field_address_exprt
+inline const field_address_exprt &to_field_address_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_field_address);
+  const field_address_exprt &ret =
+    static_cast<const field_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \copydoc to_field_address_expr(const exprt &)
+inline field_address_exprt &to_field_address_expr(exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_field_address);
+  field_address_exprt &ret = static_cast<field_address_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \brief Operator to return the address of an array element
+/// relative to a base address
+class element_address_exprt : public binary_exprt
+{
+public:
+  /// constructor for element addresses. The base address must have array type.
+  element_address_exprt(exprt base, exprt index);
```

Comment:
Hmm, `field_address_exprt` suggests that the `base` must be a pointer, but here this isn't the case?

Reply:
The `to_pointer_type(base.type())` does enforce this.

Reply:
Ok, but then the comment is inaccurate?

Reply:
Yes, fixed the comment.